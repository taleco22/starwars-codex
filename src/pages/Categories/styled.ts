import styled from "styled-components"
import {
  globalButton,
  globalItem,
  globalList,
  globalWindow,
} from "../../global-styled/index"

export const Window = styled(globalWindow)``
export const List = styled(globalList)``
export const Item = styled(globalItem)``
export const Button = styled(globalButton)``
