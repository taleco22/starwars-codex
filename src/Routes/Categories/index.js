import React from "react";
import * as S from "../../styles/styled";

export default function Categories() {
  let actual = 0;
  let anchor;

  window.onload = () => {
    anchor = document.querySelectorAll("a");
    document.addEventListener("keyup", listenKeyPress);
    anchor[0].focus();
  };

  function listenKeyPress(e) {
    if (e.keyCode === 38 && actual > 0) {
      actual -= 1;
      anchor[actual].focus();
    }
    if (e.keyCode === 40 && actual < anchor.length - 1) {
      actual += 1;
      anchor[actual].focus();
    }
  }

  return (
    <S.Window>
      <S.List>
        <S.Item>
          <S.Button autoFocus to="/categories/characters">
            Characters
          </S.Button>
        </S.Item>
        <S.Item>
          <S.Button to="/categories/planets">Planets</S.Button>
        </S.Item>
        <S.Item>
          <S.Button to="/categories/starships">Starships</S.Button>
        </S.Item>
        <S.Item>
          <S.Button to="/categories/vehicles">Vehicles</S.Button>
        </S.Item>
        <S.Item>
          <S.Button to="/categories/films">Films</S.Button>
        </S.Item>
        <S.Item>
          <S.Button to="/categories/species">Species</S.Button>
        </S.Item>
        <S.Item>
          <S.Button to="/">Go Back</S.Button>
        </S.Item>
      </S.List>
    </S.Window>
  );
}
