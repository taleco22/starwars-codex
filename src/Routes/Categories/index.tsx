import React, { useEffect } from "react";
import { useHistory } from "react-router-dom";
import * as S from "../../styles/styled";
import { initNavigation } from "../../utils/navigation";

const Categories: React.FC = () => {
  type URL = "people" | "planets" | "starships" | "vehicles" | "films" | "species";

  useEffect(() => {
    initNavigation();
  }, []);

  const history = useHistory();

  const goto = (url: URL) => {
    history.push(`/categories/${url}`);
  };

  return (
    <S.Window>
      <S.List>
        <S.Item>
          <S.Button onClick={() => goto("people")}>Characters</S.Button>
        </S.Item>
        <S.Item>
          <S.Button onClick={() => goto("planets")}>Planets</S.Button>
        </S.Item>
        <S.Item>
          <S.Button onClick={() => goto("starships")}>Starships</S.Button>
        </S.Item>
        <S.Item>
          <S.Button onClick={() => goto("vehicles")}>Vehicles</S.Button>
        </S.Item>
        <S.Item>
          <S.Button onClick={() => goto("films")}>Films</S.Button>
        </S.Item>
        <S.Item>
          <S.Button onClick={() => goto("species")}>Species</S.Button>
        </S.Item>
        <S.Item>
          <S.Button off onClick={() => history.goBack()}>
            Go Back
          </S.Button>
        </S.Item>
      </S.List>
    </S.Window>
  );
};
export default Categories;
