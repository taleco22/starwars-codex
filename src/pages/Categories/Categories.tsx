import React, { useEffect } from "react"
import { RouteComponentProps } from "react-router-dom"
import { waitButtons } from "~/utils"
import { URL } from "../types"
import { Button, Item, List, Window } from "./styled"

const Categories: React.FC<RouteComponentProps> = ({ history }) => {
  useEffect(() => {
    waitButtons()
  }, [])

  const goto = (url: URL) => {
    history.push(`/categories/${url}`)
  }

  return (
    <Window>
      <List>
        <Item>
          <Button onClick={() => goto("people")}>Characters</Button>
        </Item>
        <Item>
          <Button onClick={() => goto("planets")}>Planets</Button>
        </Item>
        <Item>
          <Button onClick={() => goto("starships")}>Starships</Button>
        </Item>
        <Item>
          <Button onClick={() => goto("vehicles")}>Vehicles</Button>
        </Item>
        <Item>
          <Button onClick={() => goto("films")}>Films</Button>
        </Item>
        <Item>
          <Button onClick={() => goto("species")}>Species</Button>
        </Item>
        <Item>
          <Button off onClick={() => history.push("/")}>
            Go Back
          </Button>
        </Item>
      </List>
    </Window>
  )
}
export default Categories
