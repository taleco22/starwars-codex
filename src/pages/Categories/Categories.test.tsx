import "@testing-library/jest-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import * as React from "react";
import Categories from "./Categories";

const use_effect = jest.spyOn(React, "useEffect");

const history = {
  push: jest.fn((url) => url),
  goBack: jest.fn((url) => url),
};

describe("Categories.tsx", () => {
  let result: { container: HTMLElement };

  beforeEach(() => {
    // @ts-ignore
    result = render(<Categories history={history} />);
  });

  it("render component correctly", () => {
    expect(result.container).toMatchSnapshot();
  });

  it("initialize component", () => {
    expect(use_effect).toBeCalled();
  });

  it("trigger onClick handler for each button", () => {
    screen.getAllByRole("button").map((v, i) => fireEvent.click(v));

    expect(history.push).toBeCalledTimes(6);
    expect(history.goBack).toBeCalledTimes(1);
  });
});
