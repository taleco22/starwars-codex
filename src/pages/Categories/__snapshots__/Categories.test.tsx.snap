// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories.tsx render component correctly 1`] = `
<div>
  <main
    class="sc-AxjAm sc-AxhUy LnXBH"
  >
    <ul
      class="sc-AxirZ sc-AxgMl eIVynF"
    >
      <li
        class="sc-AxiKw sc-AxheI bUcjaj"
      >
        <button
          class="sc-AxhCb sc-Axmtr gFtPZS"
        >
          Characters
        </button>
      </li>
      <li
        class="sc-AxiKw sc-AxheI bUcjaj"
      >
        <button
          class="sc-AxhCb sc-Axmtr gFtPZS"
        >
          Planets
        </button>
      </li>
      <li
        class="sc-AxiKw sc-AxheI bUcjaj"
      >
        <button
          class="sc-AxhCb sc-Axmtr gFtPZS"
        >
          Starships
        </button>
      </li>
      <li
        class="sc-AxiKw sc-AxheI bUcjaj"
      >
        <button
          class="sc-AxhCb sc-Axmtr gFtPZS"
        >
          Vehicles
        </button>
      </li>
      <li
        class="sc-AxiKw sc-AxheI bUcjaj"
      >
        <button
          class="sc-AxhCb sc-Axmtr gFtPZS"
        >
          Films
        </button>
      </li>
      <li
        class="sc-AxiKw sc-AxheI bUcjaj"
      >
        <button
          class="sc-AxhCb sc-Axmtr gFtPZS"
        >
          Species
        </button>
      </li>
      <li
        class="sc-AxiKw sc-AxheI bUcjaj"
      >
        <button
          class="sc-AxhCb sc-Axmtr gyNiSL"
        >
          Go Back
        </button>
      </li>
    </ul>
  </main>
</div>
`;
